contracts/verifier.sol

```diff
@@ -0,0 +1,260 @@
+        assert(isFirstStep || haveMedHash);
+        // ^ invariant if the step has been set but we don't have a hash for it
+
+        if (step == s.lowStep && step + 1 == s.medStep) {
+            // final step of the binary search (lower end)
+            s.highHash = s.medHash;
+            s.highStep = step + 1;
+        } else if (step == s.medStep && step + 1 == s.highStep) {
+            // final step of the binary search (upper end)
+            s.lowHash = s.medHash;
+            s.lowStep = step;
+        } else {
+            // this next step must be in the correct range
+            require(step > s.lowStep && step < s.highStep);
+
+            // if this is NOT the first query, update the steps and assign the correct hash
+            // (if this IS the first query, we just want to initialize medStep and medHash)
+            if (!isFirstStep) {
+                if (step < s.medStep) {
+                    // if we're iterating lower,
+                    //   the new highest is the current middle
+                    s.highStep = s.medStep;
+                    s.highHash = s.medHash;
+                } else if (step > s.medStep) {
+                    // if we're iterating upwards,
+                    //   the new lowest is the current middle
+                    s.lowStep = s.medStep;
+                    s.lowHash = s.medHash;
+                } else {
+                    // and if we're requesting the midStep that we've already requested,
+                    //   there's nothing to do.
+                    // @TODO(shrugs) - should this revert?
+                }
+            }
+
+            s.medStep = step;
+            s.medHash = bytes32(0);
+        }
+        s.lastChallengerMessage = now;
+        NewQuery(sessionId);
+    }
+
+    function respond(uint sessionId, uint step, bytes32 hash)
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        // Require step to avoid replay problems
+        require(step == s.medStep);
+        if (hash == 0) {
+            // Special "fold" signal
+            // @TODO(shrugs) - is this correct? should the claimant not fold here?
+            challengerConvicted(sessionId);
+            return;
+        }
+
+        // record the claimed hash
+        s.medHash = hash;
+        s.lastClaimantMessage = now;
+
+        // notify watchers
+        NewResponse(sessionId);
+    }
+
+    function performStepVerification(
+        uint sessionId,
+        bytes preValue,
+        bytes postValue,
+        bytes proofs,
+        address claimManager
+    )
+        onlyClaimant(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+        require(s.lowStep + 1 == s.highStep);
+        // ^ must be at the end of the binary search according to the smart contract
+
+        if (keccak256(preValue) != s.lowHash) { claimantConvicted(sessionId); return; }
```

Comment:
Probabyl better to join these two conditions.

Reply:
And we could change this into a require

Reply:
The challenger could attack by creating a micro-fork with two different med hashes, so this actually has to be a require.

Reply:
Implemented in e25f35d15eb53f48215135234e268fddaff14fcc